Service/Product/FacetBuilder.php

```diff
@@ -0,0 +1,242 @@
+    {
+        if (isset($this->facets[$storeId])) {
+            return $this->facets[$storeId];
+        }
+
+        $rawFacets = [];
+        $configFacets = $this->configHelper->getFacets($storeId);
+        foreach ($configFacets as $configFacet) {
+            if ($configFacet[self::FACET_KEY_ATTRIBUTE_NAME] === self::FACET_ATTRIBUTE_PRICE) {
+                $rawFacets = array_merge($rawFacets, array_map(
+                    function(string $attribute) {
+                        return $this->getRawFacet($attribute);
+                    },
+                    $this->getPricingAttributes($storeId)
+                ));
+            } else {
+                $rawFacets[] = $configFacet;
+            }
+        }
+
+        $this->facets[$storeId] = $this->addCategoryFacets($storeId, $rawFacets);
+        return $this->facets[$storeId];
+    }
+
+    /**
+     * @param int $storeId
+     * @param array<array<string, mixed>> $facets
+     * @return array<array<string, mixed>>
+     */
+    protected function addCategoryFacets(int $storeId, array $facets): array
+    {
+        if ($this->configHelper->replaceCategories($storeId)
+            && !array_filter($facets, function($facet) {
+                return $facet['attribute'] === self::FACET_ATTRIBUTE_CATEGORIES;
+            })
+        ) {
+            $facets[] = $this->getRawFacet(self::FACET_ATTRIBUTE_CATEGORIES);
+        }
+
+        // Added for legacy merchandising features
+        $facets[] = $this->getRawFacet(self::FACET_ATTRIBUTES_CATEGORY_ID);
+
+        if ($this->configHelper->isVisualMerchEnabled($storeId)) {
+            // Must be searchable per https://www.algolia.com/doc/guides/solutions/ecommerce/browse/tutorials/category-pages/#configure-your-index
+            $facets[] = $this->getRawFacet($this->configHelper->getCategoryPageIdAttributeName($storeId), true);
+        }
+
+        return $facets;
+    }
+
+    /**
+     * @param int $storeId
+     * @return string[]
+     * @throws NoSuchEntityException
+     * @throws LocalizedException
+     */
+    protected function getPricingAttributes(int $storeId): array
+    {
+        $pricingAttributes = [];
+        $currencies = $this->currencyManager->getConfigAllowCurrencies();
+        foreach ($currencies as $currencyCode) {
+            $pricingAttributes[] = self::FACET_ATTRIBUTE_PRICE . '.' . $currencyCode . '.default';
+            $pricingAttributes = array_merge($pricingAttributes, $this->getGroupPricingAttributes($storeId, $currencyCode));
+        }
+
+        return $pricingAttributes;
+    }
+
+    /**
+     * @param int $storeId
+     * @param string $currencyCode
+     * @return string[]
+     * @throws NoSuchEntityException
+     * @throws LocalizedException
+     */
+    protected function getGroupPricingAttributes(int $storeId, string $currencyCode): array
+    {
+        $groupPricingAttributes = [];
+        $websiteId = (int) $this->storeManager->getStore($storeId)->getWebsiteId();
```

Comment:
Micro improvement here, but can it be under the if condition as well ?

Reply:
Good call!